### PR TITLE
Added support for `mgcv::gam()` models

### LINKED ISSUE #82: Support GAMs
Making this issue so I don't forget about it. I can work on this at some point in the future

``` r
library(gtsummary)

tidy_gam <- function(x, conf.int = FALSE, exponentiate = FALSE, ...) {
  broom::tidy(x, 
                    conf.int = conf.int, 
                    exponentiate = exponentiate, 
                    parametric = TRUE,  ...) %>%
    dplyr::mutate(parametric = TRUE) %>%
    dplyr::bind_rows(
      broom::tidy(x, parametric = FALSE, ...) %>%
        dplyr::mutate(parametric = FALSE)
    ) %>%
    dplyr::relocate(.data$parametric, .after = dplyr::last_col())
}

tidy <-
  mgcv::gam(age ~ s(marker, ttdeath) + grade, data = trial) %>%
  tidy_gam(conf.int = TRUE)
tidy
#> # A tibble: 4 x 10
#>   term  estimate std.error statistic  p.value conf.low conf.high   edf ref.df
#>   <chr>    <dbl>     <dbl>     <dbl>    <dbl>    <dbl>     <dbl> <dbl>  <dbl>
#> 1 (Int~   46.2        1.85    24.9   2.37e-59    42.5      49.8    NA     NA 
#> 2 grad~    0.467      2.72     0.172 8.64e- 1    -4.86      5.79   NA     NA 
#> 3 grad~    2.13       2.69     0.792 4.30e- 1    -3.14      7.40   NA     NA 
#> 4 s(ma~   NA         NA        0.163 8.49e- 1    NA        NA       2.     2.
#> # ... with 1 more variable: parametric <lgl>

# parametric terms are tidied just like a regular model
# the smooth non-parametric terms are taken as is
# this is what I would like to see
tibble::tribble(
  ~variable, ~term, ~label, 
  "(Intercept)", "(Intercept)", "(Intercept)",
  "grade", "gradeI", "I",
  "grade", "gradeII", "II",
  "grade", "gradeIII", "III",
  "s(marker,ttdeath)", "s(marker,ttdeath)", "s(marker,ttdeath)"
) %>%
  dplyr::left_join(tidy, by = "term")
#> # A tibble: 5 x 12
#>   variable term  label estimate std.error statistic   p.value conf.low conf.high
#>   <chr>    <chr> <chr>    <dbl>     <dbl>     <dbl>     <dbl>    <dbl>     <dbl>
#> 1 (Interc~ (Int~ (Int~   46.2        1.85    24.9    2.37e-59    42.5      49.8 
#> 2 grade    grad~ I       NA         NA       NA     NA           NA        NA   
#> 3 grade    grad~ II       0.467      2.72     0.172  8.64e- 1    -4.86      5.79
#> 4 grade    grad~ III      2.13       2.69     0.792  4.30e- 1    -3.14      7.40
#> 5 s(marke~ s(ma~ s(ma~   NA         NA        0.163  8.49e- 1    NA        NA   
#> # ... with 3 more variables: edf <dbl>, ref.df <dbl>, parametric <lgl>
```

<sup>Created on 2021-01-08 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### PR DESCRIPTION
**Pull request by ddsjoberg**
closes #82  

I need to do more checks, and add unit tests.

- [ ] Add unit tests
- [ ] n_obs n_event are not populated for the smooth terms
- [ ] The default tidying function only returns the smooth terms. You need to add a custom function to include both. Is that ok?

``` r
library(gtsummary)
library(broom.helpers)

tidy_gam <- function(x, conf.int = FALSE, exponentiate = FALSE, ...) {
  broom::tidy(x, 
              conf.int = conf.int, 
              exponentiate = exponentiate, 
              parametric = TRUE,  ...) %>%
    dplyr::mutate(parametric = TRUE) %>%
    dplyr::bind_rows(
      broom::tidy(x, parametric = FALSE, ...) %>%
        dplyr::mutate(parametric = FALSE)
    ) %>%
    dplyr::relocate(.data$parametric, .after = dplyr::last_col())
}

mod <- mgcv::gam(response ~ s(marker, ttdeath) + grade, data = trial, family = binomial) 

tidy_and_attach(mod, tidy_fun = tidy_gam) %>%
  tidy_identify_variables() %>%
  tidy_add_contrasts() %>%
  tidy_add_reference_rows() %>%
  tidy_add_variable_labels() %>%
  tidy_add_term_labels() %>%
  tidy_add_header_rows() %>%
  tidy_add_n() %>%
  tidy_remove_intercept() 

tidy_plus_plus(mod, tidy_fun = tidy_gam, add_header_rows = TRUE, 
               add_estimate_to_reference_rows = TRUE)

tbl_regression(mod, tidy_fun = tidy_gam) %>%
  as_kable()
```

<table>
<thead>
<tr class="header">
<th style="text-align: left;"><strong>Characteristic</strong></th>
<th style="text-align: left;"><strong>log(OR)</strong></th>
<th style="text-align: left;"><strong>95% CI</strong></th>
<th style="text-align: left;"><strong>p-value</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Grade</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">I</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">II</td>
<td style="text-align: left;">0.26</td>
<td style="text-align: left;">-0.54, 1.1</td>
<td style="text-align: left;">0.5</td>
</tr>
<tr class="even">
<td style="text-align: left;">III</td>
<td style="text-align: left;">0.23</td>
<td style="text-align: left;">-0.56, 1.0</td>
<td style="text-align: left;">0.6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">s(marker,ttdeath)</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">0.006</td>
</tr>
</tbody>
</table>

<sup>Created on 2021-01-08 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### CODE REVIEW

**Comment by larmarange** [comment]
Thanks. Let me know when it will be ready for review.

**Comment by ddsjoberg** [comment]
@larmarange can you please take a look at this PR? I added support for mgcv GAM models, and I believe I've addressed the special cases needed.

File: R/model_get_n.R [hunk 1]
```diff
@@ -125,6 +125,34 @@ model_get_n.glm <- function(model) {
   n
 }
 
+#' @export
+#' @rdname model_get_n
+model_get_n.gam <- function(model) {
+  # get Ns from all terms
+  df_paramtric_n <- model_get_n.glm(model)
```
Comment (larmarange) [inline R/model_get_n.R:132]: typo? should it be `df_parametric_n`?

Comment (ddsjoberg) [reply R/model_get_n.R]: oops! yes!  just pushed correction

**Review by larmarange — COMMENTED** [review]
What should be `var_type` for such variables? "continuous" or something else.

Should we populate var_label or are you happy with just term name?

**Comment by ddsjoberg** [comment]
> What should be `var_type` for such variables? "continuous" or something else.

In the end it is a spline-like term, which is continuous.  I don't think it matters much what we call it, and keeping it in the family of "continuous" seemed good so that any other function that expects continuous variables to appear on a single line will correctly identify this variable on a single line

> Should we populate var_label or are you happy with just term name?

The terms can be very complex, e.g. the `s()` function has other arguments and can take many variables as well. I think just leaving it as the term will be ok, at least we'll know what to expect there rather than trying to parse something from it.

**Comment by larmarange** [comment]
Ok thanks.

Good for me. It's merged. Thanks a lot

**Comment by larmarange** [comment]
Oups, maybe a liitle too quick. 

Should we mention something in NEWS? And indicate support in the vignette.

By the way, is `tidy_gam()` available somewhere? in `gtsummary`?

Should we mention http://www.danieldsjoberg.com/gtsummary/reference/custom_tidiers.html in the documentation of `broom.helpers`?

**Comment by ddsjoberg** [comment]
> Should we mention something in NEWS? And indicate support in the vignette.
Oh yes! Added it to a housekeeping pull request.

> By the way, is `tidy_gam()` available somewhere? in `gtsummary`?
I am planning on adding it now, and I referenced the fn in the vignette.

**Pull request merged**